File: Merge_Sort.cpp
//C++ merge sort practice 

#include<iostream>
#include<vector>

//using namespace std;

void sortArray(std::vector<int>& arrayIn);
void mergeArrays(std::vector<int>& leftSub, std::vector<int>& rightSub, std::vector<int>& mergedArray);
void printArray(std::vector<int>& sortedArray);

int main(){

	std::vector<int> myArray{2,5,9,6,10,22,1,15};

	sortArray(myArray);
	std::cout<<"Sorted Array: ";
	printArray(myArray);
	
	return 0;
}

// Divide array into single element arrays 
// Top-down 
void sortArray(std::vector<int> &arrayIn){
	// if array has one element, it is sorted
	if (arrayIn.size() == 1){
		return;
	}

	std::vector<int> leftSub;
	std::vector<int> rightSub;
	int mid = arrayIn.size()/2;

	// Put the first half of the array into array one
	for (int i=0; i<arrayIn.size()/2; ++i){
		leftSub.push_back(arrayIn[i]);
	}
	// Put the second half of the array into array two
	for (int i=mid; i<arrayIn.size(); ++i){
		rightSub.push_back(arrayIn[i]);
	}

	// Use recursion to continue to split arrays until the baseline condition is met
	sortArray(leftSub);
	sortArray(rightSub);
	mergeArrays(leftSub, rightSub, arrayIn);

}


// Merge array elements into smaller sorted arrats
// Bottom-up portion
//void mergeArrays(std::vector<int>& leftSub, std::vector<int>& rightSub){
void mergeArrays(std::vector<int> &leftSub, std::vector<int> &rightSub, std::vector<int> &mergedArray){
	
	mergedArray.clear();
	
	while ( (!leftSub.empty()) && (!rightSub.empty()) ){
		if (leftSub[0] <= rightSub[0]){
			mergedArray.push_back(leftSub[0]);
			leftSub.erase(leftSub.begin());
		}
		else {
			mergedArray.push_back(rightSub[0]);
			rightSub.erase(rightSub.begin());
		}
	}
	
	// Either arrayA or arrayB are empty
	// Add remaining sorted values from A or B into merged array
	while (!leftSub.empty()){
		mergedArray.push_back(leftSub[0]);
		leftSub.erase(leftSub.begin());
	}

	while (!rightSub.empty()){
		mergedArray.push_back(rightSub[0]);
		rightSub.erase(rightSub.begin());
	}
	
}


void printArray(std::vector<int> &sortedArray){
	std::cout<<"{ ";
	for (int i=0; i<sortedArray.size(); ++i){
		std::cout<<sortedArray[i]<<" ";
	}
	std::cout<<"}"<<"\n";
}

File: Merge_Sort.py
#!/usr/bin/python3

# Merge sort practice 

def main():
	sort_array(myArray)
	print("Sorted Array: %s" % myArray)
	

def sort_array(sorted_array):
	#recursive function to split array 
	#baseline
	if len(sorted_array) > 1:
	
		mid = len(sorted_array)//2
		left_array = sorted_array[:mid]
		right_array = sorted_array[mid:]

		sort_array(left_array)
		sort_array(right_array)

		merge_arrays(left_array, right_array, sorted_array)


def merge_arrays(left_array, right_array, sorted_array):
	#merge into sorted array 
	sorted_array.clear() 

	while (left_array and right_array):
		if (left_array[0] <= right_array[0]):
			sorted_array.append(left_array[0])
			left_array.pop(0)
		else:
			sorted_array.append(right_array[0])
			right_array.pop(0)

	# Add items from remaining sorted list
	while (left_array):
		sorted_array.append(left_array[0])
		left_array.pop(0)

	while (right_array):
		sorted_array.append(right_array[0])
		right_array.pop(0)


if __name__ == '__main__':
	myArray = [2,6,8,5,91,52,11,25]
	main()
